agent/src/beerocks/slave/backhaul_manager/backhaul_manager_thread.cpp

```diff
@@ -3411,32 +3397,46 @@ bool backhaul_manager::handle_1905_topology_discovery(const std::string &src_mac
 
     auto mid = cmdu_rx.getMessageId();
     LOG(INFO) << "Received TOPOLOGY_DISCOVERY_MESSAGE from AL MAC=" << tlvAlMac->mac()
-              << ", mid=" << std::hex << int(mid);
+              << ", mid=" << std::hex << mid;
 
     auto tlvMac = cmdu_rx.getClass<ieee1905_1::tlvMacAddress>();
     if (!tlvMac) {
         LOG(ERROR) << "getClass tlvMacAddress failed";
         return false;
     }
 
-    uint32_t if_index   = message_com::get_uds_header(cmdu_rx)->if_index;
-    std::string if_name = network_utils::linux_get_iface_name(if_index);
-    if (if_name.empty()) {
+    uint32_t if_index                      = message_com::get_uds_header(cmdu_rx)->if_index;
+    std::string local_receiving_iface_name = network_utils::linux_get_iface_name(if_index);
+    if (local_receiving_iface_name.empty()) {
         LOG(ERROR) << "Failed getting interface name for index: " << if_index;
         return false;
     }
 
-    auto new_device =
-        m_1905_neighbor_devices.find(tlvAlMac->mac()) == m_1905_neighbor_devices.end();
+    std::string local_receiving_iface_mac_str;
+    if (!network_utils::linux_iface_get_mac(local_receiving_iface_name,
+                                            local_receiving_iface_mac_str)) {
+        LOG(ERROR) << "Failed getting MAC address for interface: " << local_receiving_iface_name;
+        return false;
+    }
+
+    // Check if it is a new device so if it does, we will send Topology Notification.
+    bool new_device = false;
+    for (auto &neighbors_on_local_iface_entry : db->neighbor_devices) {
+        auto &neighbors_on_local_iface = neighbors_on_local_iface_entry.second;
+        new_device =
+            neighbors_on_local_iface.find(tlvAlMac->mac()) == neighbors_on_local_iface.end();
+        if (new_device) {
+            break;
+        }
+    }
```

Comment:
this entire check can move down closer to where the new device is added and instead just handle the case (send the message).
even if you want to do it in 2 steps - which is ok from a readability perspective, you can still move it down not to repeat the explanation regarding new-device

Reply:
We can't do that. If we did, we will always get it is not a new device since we added it to the list right after it.